source/_posts/react-localstorage.md

```diff
@@ -0,0 +1,382 @@
+    localStorage.setItem('counter', count)
+  }, [count])
+
+  return (
+    <div className="card">
+      <h3>Medium Example</h3>
+      <p>Click the button many times and try to refresh the page</p>
+      <br />
+      <button onClick={handleCounter}>
+        count is {count}
+      </button>
+    </div>
+  ) 
+}
+```
+
+{% img "medium_example.png" "Medium react component, web storage" "lazy" %}
+
+Now, try to refresh the page, and the counter state will be the same. But imagine if you should do the same in multiple components, that is terrible! But don't worry. We have a solution, React Hooks.
+
+### Implementing a local storage hook
+
+React Hooks allows to reuse of the code and extends the `localStorage` functionality to other components. Let's see the implementation after we need a dedicated folder to abstract the hook, something like this.
+
+{% img "folder_structure.png" "react project structure, src folder" "lazy" %}
+
+To create a custom hook, we only need a function that returns a state and its set function, as we see in the following code (`localstorage.js`, our custom hook).
+
+```js
+import { useEffect, useState } from "react"
+
+const decode = (value) => {
+  return JSON.stringify(value)
+}
+
+const encode = (value) => {
+  return JSON.parse(value)
+}
+
+// useLocalStorage hook
+const useLocalStorage = (key, defaultState) => {
+  const [value, setValue] = useState(
+    encode(localStorage.getItem(key)||null) || defaultState
+  )
+
+  useEffect(() => {
+    localStorage.setItem(key, decode(value))
+  },  [value])
+
+  return [value, setValue]
+}
+
+export {
+  useLocalStorage
+}
+```
+
+Now, every component only needs to use a unique name key to avoid colliding with key-value pairs of the others and a default state as a second parameter. It is a simple and helpful implementation, but if you need something mature and complete, we recommend finding open-source libraries like [use-local-storage-state](https://github.com/astoilkov/use-local-storage-state).
+
+So, we have a new custom React hook called `useLocalStorage` to replace the previous implementation.
+
+```js
+import { useLocalStorage } from "../hooks/localstorage"
+
+export default function HookMode(){
+  const [counter, setCounter] = useLocalStorage('counter', 0);
+  const [secondCounter, setSecondCounter] = useLocalStorage('secondCounter', 0);
+  
+  return (
+    <div className="card">
+      <h3>Hook Example</h3>
+      <p>Click the button many times and try to refresh the page</p>
+      <br />
+
+      <button onClick={() => setCounter(counter + 1)}>
+        Button 1: count is {counter}
+      </button>
+      <br /><br />
+      <button onClick={() => setSecondCounter(secondCounter + 1)}>
+        Button 2: count is {secondCounter}
+      </button>
+    </div>
+  )
+}
+```
+
+{% img "hook_example.png" "Hook react component, react localstorage, hook style" "lazy" %}
+
+See how the code reduces a lot and how we can reuse the hook (many times as we need).
```

Comment:
"See how the code reduces a lot" hmm does not sound good grammar-wise :D code shrinks (itself) maybe rather reduces (something - what?) ?

Reply:
Or _the amound code IS reduced_ ?

Reply:
Done :D